### ISSUE TITLE
Comparison to dense_vector and cosineSimilarity

### ISSUE THREAD

**Issue by Querela** [StaySense/fast-cosine-similarity#12]
Hi,
I really like your plugin. I had to update to the latest version (7.0.1) for my usecase and it still worked! (_I later incorporated andrassy's changes when I realized that there are some current forks..._)
**I found a builtin feature that is really similar, so I want to ask how they compare.**

- [**dense_vector** mapping type](https://www.elastic.co/guide/en/elasticsearch/reference/master/dense-vector.html) (as well as a sparse variant)
- [cosineSimilarity, dotProduct](https://www.elastic.co/guide/en/elasticsearch/reference/master/query-dsl-script-score-query.html#vector-functions)  
  &rarr; **only in master/future version** (not in current 7.0.1)  
  because I either do something wrong or it is not implemented (according to docs), as I always get a `compile error` in der `search_phase_execution_exception` when searching

The limitation seems to be 500 (current) / 1024 (master/future) Elements for dense_vector. And it is an experimental feature but I'm interested in the performance.
For 5 Mio. doc2vec (dim=300) vectors I currently have an average search time of about 5 seconds. (Retrieval of search-vector for given id included.)

**Comment by wmelton** [comment]
Hi @Querela - 

"For 5 Mio. doc2vec (dim=300) vectors I currently have an average search time of about 5 seconds. " You're saying that performance number is using our plugin or the native dense_vector mapping type?

**Comment by Querela** [comment]
Hi @wmelton 

**I'm currently using the StaySense plugin.** I first had to do the negative adjustment because it threw errors (see my commit history) before I found/saw that there were other forks ... Currently, I'm using the andrassy version for 7.0.1 using the pom, because I only have java 11.
There were not really noticable differences between the original version and the version from andrassy (as far as I could tell). I did not do serious performance tests. (I also was using Python.)

The performance is most probably depending on my disk performance, I'm not so sure about caching. I only assigned max 4GB (-xms/xmx) for the java process even if htop displays almost the same as the index size. Not that sure how it works behind the scences.

I tried the native dense_vector type (and the index was between 1,5 - 2 times larger than the plugin index). I did not try to modify the plugin source to use the `dense_vector` field because I'm not that clear about the source code and inner workings of elasticsearch ... (not that much docs for this type)
As far as I could tell, the `cosineSimilarity` appears in version 7.2.0 (?), see [github sources](https://github.com/elastic/elasticsearch/search?q=cosineSimilarity&unscoped_q=cosineSimilarity).

I also tried to use scripts (painless) to do a simple dot product but it would not work for me. All variants of the following snippet, (with/without function_score etc. etc.):
```python
{
    "query": {
        "function_score": {
            "boost_mode" : "replace",
            "functions": [
                {
                    "script_score": {
                        # "query": {
                        #     "match_all": {}
                        # },
                        "script": {
                            "lang": "painless",
                            "source": """
                              Debug.explain(doc);
                              double score = 0.0d;
                              Debug.explain(doc['{docVec}']);
                              for (int i = 0; i < params.query_vector.length; ++i) {{
                                score += doc['{docVec}'][i] * params.query_vector[i];
                              }}
                              return sigmoid(1, Math.E, -score);
                            """.format(docVec=ES_INA_VEC),  # my mapping field name
                            "params": {
                                "query_vector": vec,
                            },  # params
                        }, # script
                    },  # script_score
                },
            ], # functions
        }, # function_score
    }, # query
}
```
The most telling message of all tries was: `Field [embeddingVector] of type [dense_vector] doesn't support sorting, scripting or aggregating`. That most probably means I either have to store the field with some additional configs (doc_values, ...?) or it is not for iterating?
I also tried debugging with `Debug.explain(doc['{docVec}']);` but this also did not really work ...

TL;DR - So ... I'm using the plugin.

**Comment by wmelton** [comment]
@Querela - 

Using painless is a really smart idea for this - I have not had a chance to do performance benchmarking on this yet but had decided that if the numbers were comparable, I might retire our plugin altogether and recommend using default functionality.

Regarding your query, I think you should try something like this: 

```

{
"query": {
        "match_all": {}
    },
	"script_fields":{
		"vectorDistance":{
			"script":{
				"lang":"painless",
				"source":"double score = 0.0d; for (int i = 0; i < params.query_vector.length; ++i) {{ score += doc['{docVec}'][i] * params.query_vector[i];}}return sigmoid(1, Math.E, -score);",
        "params":{
          "query_vec": vec
        }
			}
		}
	}
}
```

This should return a calculated field called 'vectorDistance' that is the dotProduct. 

I haven't tested this, but it looked like there could be some syntax errors. 

Additionally, if that doesn't work - you may want to try building an index with a field **not** defined as a dense_vector to see if the field definition is causing some invisible failure we don't have visibility to.

**Comment by Querela** [comment]
@wmelton 
Yes, I thought, painless should _maybe_ better but I was just starting with elasticsearch (my use case was on-the-fly k-nearest-neighbor search on limited RAM), so I'm more or less stumbling around ...
I'm not sure what influences what, e. g. how painless caches nested variables, can I define functions and call them, how does the query DSL work (function score, script score ...) etc.

This way (`"match_all": {}`) may be better to run kNN/cosine similarity on a limited subset. I was not really sure how queries are concatenated (select subset, run "subquery" on the subset etc.)

Running the test may take some hours, to set it up ... I can also dump my setup as a gist?

**Comment by Querela** [comment]
Found this:

https://github.com/elastic/elasticsearch/issues/31615
> The indexing of vectors are available from v7.0.0-beta1, but querying of them will be avaialable only from v7.1.

https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-script-fields.html
See last paragraph (I may have to change the access to `params['_source']['{docVec}']` which is slower:
> Also, the doc[...] notation only allows for simple valued fields (you can’t return a json object from it) and makes sense only for non-analyzed or single term based fields.

The easiest would probably be to wait for the 7.1.* (?) version when official support is implemented. For older versions the StaySense plugin.

**Comment by wmelton** [comment]
@Querela - in the interim, you should be able to just store the scalar value in an Elasticsearch field and use Painless to test. 

I would definitely stay away from using _source values as you are correct, performance is definitely going to suffer greatly.

**Comment by Querela** [comment]
@wmelton 

Initial results:
```java
java.lang.UnsupportedOperationException: Field [embeddedVector] of type [dense_vector] doesn't support sorting, scripting or aggregating
```
Had to change to `return (score + 1.0d) * 0.5d;`, since `sigmoid` was unknown ...

It worked for:
```python
req_body = {
    "query": {
        "match_all": {},
    },
    "script_fields": {
        "vectorDistance": {
            "script": {
                "lang": "painless",
                # "source": "double score = 0.0d; for (int i = 0; i < params.query_vector.length; ++i) {{ score += doc['{docVec}'][i] * params.query_vector[i]; }} return sigmoid(1, Math.E, -score);".format(docVec=ES_INA_VEC),
                # "source": "double score = 0.0d; for (int i = 0; i < params.query_vector.length; ++i) {{ score += doc['{docVec}'][i] * params.query_vector[i]; }} return (score + 1.0d) * 0.5d;".format(docVec=ES_INA_VEC),
                # "source": "double score = 0.0d; double[] docVec = doc['{docVec}']; for (int i = 0; i < params.query_vector.length; ++i) {{ score += docVec[i] * params.query_vector[i]; }} return (score + 1.0d) * 0.5d;".format(docVec=ES_INA_VEC),
                "source": "double score = 0.0d; for (int i = 0; i < params.query_vector.length; ++i) {{ score += params['_source']['{docVec}'][i] * params.query_vector[i]; }} return (score + 1.0d) * 0.5d;".format(docVec=ES_INA_VEC),
                "params": {
                    "query_vector": vec,
                }, # params
            }, # script
        }, # vectorDistance
    }, # script_fields
} # query
```

I have to access it via: `params['_source']['{docVec}'][i]`

Now I guess I have to look, how I use the script_score fields to rank my docs.

**Comment by wmelton** [comment]
@Querela - I think you need to stop using the `dense_vector` field definition until 7.1 is released. Change your 'embeddedVector' definition to `array` and store it as an array of values e.g. `[1.0, .04, ..., n]`. 

This way you shouldn't need to use `_source`. Using `_source` is going to degrade performance to make the exercise useless.

**Comment by Querela** [comment]
@wmelton 
In this case, we can just use the plugin. The idea was to _maybe_ improve performance with the use of `dense_vector` and `painless` scripting. Using an array would have already been considered when writing the plugin, wouldn't it? I'm also not sure how array would react. In the docs it looked like it might reorder the elements?

When I replaced the working script/source into the old `function_score` query it simply resulted in a timeout error.

Therefore, staying with the plugin until the new version. It just works, is more or less fast enough for my use-case. I have to research if I can improve it with pre-filtering but I'm currently not that sure how ....

I would close the issue if there is nothing more?

**Comment by wmelton** [comment]
@Querela - reordering the data behind the scenes is a good consideration. My only purpose for suggesting the changes was to see if there were any obvious performance improvements using painles vs our plugin. 

One option would be to denormalize the vectors in to json docs as so: 

```
{
  'value0' : 0.34,
  'value1' : 1.1,
  'valuen' : -0.01
}
```

And then use painless to compute each index of the query vector against the associated doc value. Again, haven't tested so performance could be worse. 

In terms of filtering, using a `must` filter in the `query` block should narrow down the documents to just the ones you care about. Using our plugin should calculate similarity and return them correctly ordered.

Thanks!

**Comment by Querela** [comment]
Reordering is a guess. I'm not sure but lists are implicit so there may be magic happening in the background to improve access. There might not be a reason for this for floats than for strings but I'm not that sure. And to build a test case is not my foremost priority at this time.

Well, denormalization would probably work. But I'm rather sceptic what the outcome could be. The index would probably explode because of the many fields. My workflow is first getting a vector for a given id and then using the vector to compute similar items. With denormalization this would be really explicit and not that intuitive. And the query (with `painless`) would either use some string concatenation for indexing (`doc['value' + x]` or have an explicit list of multiplications ...
I guess denormalization has to be worse somehow or else there would be no (reason for a) `dense_vector` type. It's kind of bad that there is no real usage until they have the implementation of `consineSimilarity` since the field doesn't really allow efficient scripting.
I guess there could be a way to use the `dense_vector` field as `doc_values` field in the plugin but I have not easily found the way to access it. (It may be to recent an addition to have any examples, and documentation is sparse.)

Thanks for the hint of `must`. I was going to test it sometime in the future but was not sure whether elasticsearch would _optimize_ the order of execution or if it would generate two subsets (from restricting and from scoring) and then combine those for the output ... I remember that I read somewhere that the execution is in two phases where the ordering happens after the filtering (?), and the scoring would probably be after the filtering. Well, have to test it out ...

I try to append my test setup with python gensim (doc2vec) and elasticsearch later today for those that may want to use this themselves or to test lists etc.

**Comment by wmelton** [comment]
@Querela - I agree with you on the `bucket explosion` problem using the denormalized version. I had never tested that personally, so it was more of a "what would happen if..." type thought. 

Using the `must` filter should most definitely save significant compute time. From all of the Elastic engineers thoughts they've shared in the forums, this seems to be the status quo for increasing query performance. Bitset filters (must) are cached differently than other query parameters allowing special query optimizations to happen.

**Comment by Querela** [comment]
Hi, didn't know about `bucket explosion` but was a conjecture (?) when comparing to other database types, JSON etc. Maybe a schema mapping might improve this a little bit? It just has to know which value corresponds to which key. In a (binary) row format with pointers it would be good but in elasticsearch each field seems kind of optional ... `doc_values` can perhaps help?

A toy example with simple lists may be good for insights into reordering.

Ok. Thanks, I still have to try it out. ;-)

**Comment by Querela** [comment]
For comparison: I tried the same vector collection on another server (with less RAM than the index size, 8 GB) and now I'm getting attrocious results. Now it takes about 65 seconds per search. (I had to use `request_timeout` in the python library.)
I have to try to filter some subset to only compute it there. That may improve the results.

I guess it is a disk performance issue. I'm not sure what types of disks I have on the servers I use but I guess this should be the deciding factor?

**Comment by wmelton** [comment]
@Querela - 65 seconds! That's awful. How large is your collection? 

Realistically, this is one of the downsides to the ElasticSearch solution they're implementing. Linear Scans for collections much bigger than 1000 vectors is going to yield terrible performance results.
